src/oauth-provider.ts

```diff
@@ -2290,19 +2293,284 @@ class OAuthProviderImpl {
   }
 
   /**
-   * Fetches client information from KV storage
+   * Fetches client information from KV storage or via CIMD (Client ID Metadata Document)
    * This method is not private because `OAuthHelpers` needs to call it. Note that since
    * `OAuthProviderImpl` is not exposed outside this module, this is still effectively
    * module-private.
+   *
+   * Supports CIMD: If clientId is an HTTPS URL with a non-root path, the metadata
+   * document will be fetched from that URL instead of looking up in KV storage.
+   *
    * @param env - Cloudflare Worker environment variables
-   * @param clientId - The client ID to look up
+   * @param clientId - The client ID to look up (can be a regular ID or an HTTPS URL for CIMD)
    * @returns The client information, or null if not found
    */
-  getClient(env: any, clientId: string): Promise<ClientInfo | null> {
+  async getClient(env: any, clientId: string): Promise<ClientInfo | null> {
+    // Check if this is a CIMD (Client ID Metadata Document) URL
+    if (this.isClientMetadataUrl(clientId)) {
+      return this.fetchClientMetadataDocument(env, clientId);
+    }
+
+    // Standard KV lookup
     const clientKey = `client:${clientId}`;
     return env.OAUTH_KV.get(clientKey, { type: 'json' });
   }
 
+  /**
+   * Checks if a client_id is a valid CIMD URL
+   * Must be HTTPS with a non-root pathname
+   */
+  private isClientMetadataUrl(clientId: string): boolean {
+    try {
+      const url = new URL(clientId);
+      return url.protocol === 'https:' && url.pathname !== '/';
```

Comment:
this is correct to reject here, but you may want a louder failure if someone's passed this situation, since i think with this it will fail kind of oddly (like will it fall back to DCR?).  i imagine it'll be rare, so fine to cross that bridge if it comes up.